### PR TITLE
[Merged by Bors] - refactor(Init.Data.List.Instances): sync to Mathlib 3

### PR DESCRIPTION
**Pull request by Komyyy**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Init/Data/List/Instances.lean [hunk 1]
```diff
@@ -2,15 +2,51 @@
...
+  (id_map := map_id)
+  (pure_bind := fun _ _ => List.append_nil _)
+  (bind_assoc := by
+    intro _ _ _ l _ _
+    induction l with
+    | nil => simp [Bind.bind]
+    | cons _ _ ih => simp [Bind.bind] at ih; simp [Bind.bind, ih])
+  (bind_pure_comp := by
+    intro _ _ _ l
+    induction l <;> simp [Bind.bind, Functor.map, pure, List.ret] at *; assumption)
```
Comment (jcommelin) [inline Mathlib/Init/Data/List/Instances.lean:39]: Can't you prove these with 1-liners like in the old code? Or do you need a heavy import for that?

Comment (Komyyy) [reply Mathlib/Init/Data/List/Instances.lean:39]: Yes. The old code is depends on lemmas in Data.List.Basic.

Comment (Komyyy) [reply Mathlib/Init/Data/List/Instances.lean:39]: Yes. The old code is depends on lemmas in Data.List.Basic.

Comment (jcommelin) [reply Mathlib/Init/Data/List/Instances.lean:39]: Could you please leave a comment explaining how to prove these in 1 line if that import is present?
We should probably split `Data.List.Basic` into smaller pieces. Then this file can have a lightweight import that is useful for proofs like these.

**Comment by jcommelin** [comment]
bors d+

File: Mathlib/Init/Data/List/Instances.lean (hunk 1 above)
Comment (Komyyy) [reply Mathlib/Init/Data/List/Instances.lean]: I moved some lemmas from `Data.List.Basic`.

**Pull request closed without merging**